src/rp2_common/hardware_uart/uart.c

```diff
@@ -69,6 +70,65 @@ void uart_deinit(uart_inst_t *uart) {
     uart_reset(uart);
 }
 
+static uint32_t uart_disable_before_lcr_write(uart_inst_t *uart) {
+    // Notes from PL011 reference manual:
+    //
+    // - Before writing LCR, must disable UART and wait for current TX + RX
+    //   char to finish
+    //
+    // - There is a BUSY flag which waits for the current TX char, but this is
+    //   OR'd with TX FIFO !FULL, so not usable when FIFOs are enabled and
+    //   potentially nonempty
+    //
+    // - FIFOs can't be set to disabled whilst a character is in progress
+    //   (else "FIFO integrity is not guaranteed")
+    //
+    // Combination of these means there is no general way to halt and poll for
+    // end of TX character, if FIFOs may be enabled. Either way, there is no
+    // way to poll for end of RX character.
+    //
+    // So... we're going to insert a 15 baud period delay before changing
+    // settings (comfortably higher than max data + start + stop + parity).
+    // Anything else would require API changes to permit a non-enabled UART
+    // state after init() where settings can be changed safely.
+
+    uint32_t cr_save = uart_get_hw(uart)->cr;
+    hw_clear_bits(&uart_get_hw(uart)->cr,
+        UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS);
+
+    uint32_t current_ibrd = uart_get_hw(uart)->ibrd;
+    uint32_t current_fbrd = uart_get_hw(uart)->fbrd;
+
+    // Note: Maximise precision here. Show working, the compiler will mop this up.
+    // Create a 16.6 fixed-point fractional division ratio; then scale to 32-bits.
+    uint32_t brdiv_ratio = 64u * current_ibrd + current_fbrd;
+    brdiv_ratio <<= 10;
+    // 3662 is ~(15 * 244.14) where 244.14 is 16e6 / 2^16
+    uint32_t scaled_freq = clock_get_hz(clk_peri) / 3662ul;
+    uint32_t wait_time_us = brdiv_ratio / scaled_freq;
```

Comment:
I'm not convinced this math is right yet (or the comments)

I calc for 133Mhz peri clock and 115200 baud rate, then brdiv_ratio should be 9236 ish.
Thhe above calc then gives me (I think) 260us

However 15 bauds is 130us

Reply:
I think this is correct as submitted.
You got to 260 us because there's a clever twist in the Baud rate setting code that divides the integer part of the divide ratio by _128_ using a `>> 7` and *not* by 64 as would be assumed from the 6-bit fractional part...
That's why the the code that calculates the divisor uses 8 and not the 16 required by the UART hardware.
`uint32_t baud_rate_div = (8 * clock_get_hz(clk_peri) / baudrate);`
I suspect that was to avoid wrapping a 32-bit number at low Baud rate with fast `clk_peri` rates.

Reply:
ah ok, thanks; i had assumed the x8 was for 8 bits ;-)